aten/src/ATen/native/cuda/Copy.cu

```diff
@@ -261,7 +261,113 @@ static void copy_kernel_cuda(TensorIterator& iter, bool non_blocking) {
   }
 }
 
+int64_t uvm_get_guard_index(Tensor& t) {
+  int cuda_device_index;
+  if (t.is_cpu()) {
+    // TODO: We can to do more here once we have
+    // indirect contexts in place
+    cuda_device_index = -1;
+  } else {
+    TORCH_CHECK(t.is_cuda());
+    cuda_device_index = t.get_device();
+  }
+  return cuda_device_index;
+}
+
+void uvm_cuda_mem_advise_preferred_location(Tensor t) {
+  // Call cudaMemAdvise on a managed tensor
+  at::cuda::OptionalCUDAGuard device_guard;
+  int64_t cuda_device_index = uvm_get_guard_index(t);
+  int hint_device;
+  if (t.is_cpu()) {
+    hint_device = -1 /*cudaCpuDeviceId*/;
+  } else {
+    hint_device = static_cast<int>(cuda_device_index);
+  }
+
+  void* ptr = t.data_ptr();
+  size_t size_bytes = at::detail::computeStorageNbytes(
+      t.sizes(), t.strides(), t.dtype().itemsize());
+
+  device_guard.set_index(cuda_device_index);
+  AT_CUDA_CHECK(cudaMemAdvise(
```

Comment:
Should we also expose other cudaMemAdvise API (https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__TYPES.html#group__CUDART__TYPES_1gc314a8b14091f7e02a7ad15dcb36c857) to the users? In FBGEMM, the API is defined as https://github.com/pytorch/FBGEMM/blob/5876514306de2586e617b6ff1212eacf25114a49/fbgemm_gpu/src/cumem_utils.cu#L364 , and the usages are https://github.com/pytorch/FBGEMM/blob/5876514306de2586e617b6ff1212eacf25114a49/fbgemm_gpu/test/uvm_test.py#L103

Reply:
yes, but at the moment, this is not exposed to the user like in FBGEMM, so we may not need all of them.  Only the "move kernel" here uses it.